tools/osv-scanner/internal/output/machinejson.go

```diff
@@ -0,0 +1,71 @@
+package output
+
+import (
+	"encoding/json"
+	"log"
+
+	"github.com/google/osv.dev/tools/osv-scanner/internal/osv"
+)
+
+type Output struct {
+	Results []Result `json:"results"`
+}
+
+type Result struct {
+	FilePath string    `json:"filePath"`
+	Packages []Package `json:"packages"`
+}
+
+type Package struct {
+	Name            string              `json:"name"`
+	Version         string              `json:"version"`
+	Ecosystem       string              `json:"ecosystem"`
+	Vulnerabilities []osv.Vulnerability `json:"vulnerabilities"`
+}
+
+func PrintJSONResults(query osv.BatchedQuery, resp *osv.HydratedBatchedResponse) {
+	output := Output{}
+	groupedBySource := map[string][]Package{}
+
+	for i, query := range query.Queries {
+		response := resp.Results[i]
+		if len(response.Vulns) == 0 {
+			continue
+		}
+		var pkg Package
+		if query.Commit != "" {
+			pkg.Version = query.Commit
+		} else if query.Package.PURL != "" {
+			var err error
+			pkg, err = PurlToPackage(query.Package.PURL)
+			if err != nil {
+				log.Printf("Failed to parse purl: %s, with error: %s",
+					query.Package.PURL, err)
+				continue
+			}
+			pkg.Vulnerabilities = response.Vulns
+		} else {
+			pkg = Package{
+				Name:            query.Package.Name,
+				Version:         query.Version,
+				Ecosystem:       query.Package.Ecosystem,
+				Vulnerabilities: response.Vulns,
+			}
+		}
+
+		groupedBySource[query.Source] = append(groupedBySource[query.Source], pkg)
+	}
+
+	for source, packages := range groupedBySource {
+		output.Results = append(output.Results, Result{
+			FilePath: source,
+			Packages: packages,
+		})
+	}
+
+	marshal, err := json.MarshalIndent(output, "", "  ")
+	if err != nil {
+		return
+	}
+	print(string(marshal))
```

Comment:
rather than do this, can we just create an `Encoder` (https://pkg.go.dev/encoding/json#Encoder) and set the output stream to `os.Stdout`.